### ISSUE TITLE
sdia psu provider not producing machine metrics

### ISSUE THREAD

**Issue by marvil07-adapt** [green-coding-berlin/green-metrics-tool#327]
## Description

Hi!
This project is quite nice, thanks for making it available :+1:

It seems like even if the sdia provider is enabled and configured with the two required parameters at `config.yml`, it is not producing an overall result for machine energy.

By enabled and configured, I mean that the following lines are part of the `config.yml`, under `measurement.metric-providers.common`.

          psu.energy.ac.sdia.machine.provider.PsuEnergyAcSdiaMachineProvider:
            resolution: 100
            CPUChips: 2
            TDP: 28

Here some of relevant the output produced by `runner.py` when being executed for the stress example application.
The command used was `python3 runner.py --uri /full/path/to/example-applications/stress --name stress`.

    ...
    Importing metric providers 
    ...
    Importing PsuEnergyAcSdiaMachineProvider from metric_providers.psu.energy.ac.sdia.machine.provider
    Configuration is {'resolution': 100, 'CPUChips': 2, 'TDP': 28}
    ...
    Waiting for Metric Providers to boot ... 
    ...
    Stderr check on PsuEnergyAcSdiaMachineProvider
    ...
     Stopping metric providers and parsing measurements 
    ...
    Imported  474  metrics from  PsuEnergyAcSdiaMachineProvider
    ...
    Killing processes  
     -Cleanup gracefully completed 
    ...
    >>>> MEASUREMENT SUCCESSFULLY COMPLETED <<<<
    ...

The front-end application reports the following on the main metrics.
![Screenshot 2023-06-15 at 11-01-20 Green Metrics Tool](https://github.com/green-coding-berlin/green-metrics-tool/assets/75342537/55adeedf-3b21-4579-9fde-c7f7b4f89621)

And the detailed information under it is the following, where there is no _Machine energy_ row.

Metric | Source | Scope | Detail Name | Value | StdDev | Type | MAX
-- | -- | -- | -- | -- | -- | -- | --
CPU Energy (Package) | RAPL | component | Package_0 | 55.54 | J | N/A |  
CPU Frequency | sysfs | core | 0 | 3.07 | GHz | N/A | 3.50 GHz (MAX)
CPU Frequency | sysfs | core | 1 | 3.07 | GHz | N/A | 3.50 GHz (MAX)
CPU Frequency | sysfs | core | 2 | 3.08 | GHz | N/A | 3.50 GHz (MAX)
CPU Frequency | sysfs | core | 3 | 3.07 | GHz | N/A | 3.50 GHz (MAX)
CPU Power (Package) | RAPL | component | Package_0 | 9.05 | W | N/A | 10.99 W (MAX)
CPU % | cgroup | container | gcb-alpine-stress | 20.28 | % | N/A | 25.64 % (MAX)
CPU % | procfs | system | [system] | 23.58 | % | N/A | 29.26 % (MAX)
CPU Temperature | lm_sensors | component | acpitz-acpi-0_temp2 | 29.00 | °C | N/A | 29.00 °C (MAX)
CPU Temperature | lm_sensors | component | ath10k_hwmon-pci-0100_temp1 | 49.45 | °C | N/A | 50.00 °C (MAX)
CPU Temperature | lm_sensors | component | coretemp-isa-0000_Core-0 | 43.10 | °C | N/A | 52.00 °C (MAX)
CPU Temperature | lm_sensors | component | coretemp-isa-0000_Core-1 | 54.31 | °C | N/A | 64.00 °C (MAX)
CPU Temperature | lm_sensors | component | coretemp-isa-0000_Package-id-0 | 54.76 | °C | N/A | 64.00 °C (MAX)
CPU Temperature | lm_sensors | component | nvme-pci-3b00_Composite | 32.00 | °C | N/A | 32.00 °C (MAX)
CPU Temperature | lm_sensors | component | nvme-pci-3b00_Sensor-1 | 32.00 | °C | N/A | 32.00 °C (MAX)
CPU Temperature | lm_sensors | component | nvme-pci-3b00_Sensor-2 | 60.14 | °C | N/A | 61.00 °C (MAX)
Memory Energy (DRAM) | RAPL | component | Package_0 | 1.83 | J | N/A |  
Memory Power (DRAM) | RAPL | component | Package_0 | 0.30 | W | N/A | 0.49 W (MAX)
Memory Usage | cgroup | container | gcb-alpine-stress | 8.10 | MB | N/A | 9.65 MB (MAX)
Network CO2 | Formula | global | [FORMULA] | 0.00 | g | N/A |  
Network Energy | Formula | global | [FORMULA] | 0.02 | J | N/A |  
Network I/O | cgroup | container | gcb-alpine-stress | 0.00 | MB | N/A |  
Phase Duration | Syscall | system | [SYSTEM] | 6.14 | s | N/A |  

## Definition of Done

sdia produces machine level energy metrics.

## Notes

Used code version was 28d656abfc56f7c6f6e99420630111bd50d810d5.

This started as a side topic at https://github.com/green-coding-berlin/example-applications/pull/45#issuecomment-1590243113.
Moving it here, where it is likely more on topic.

Any pointers will be appreciated.
BTW if I should post this elsewhere instead of here, please let me know.

**Comment by ArneTR** [comment]
Thanks for taking the time to make a detailed write up. Now I understand the issue better!

The problem is some legacy issue with the GMT. In order to utilize the model based metrics some CPU utilization metric must be fed in. 

We currently have no reporter that could provide that for macOS. For linux the `cpu.utilization.procfs.system.provider.CpuUtilizationProcfsSystemProvider` reporter is used to generate that data. It acutally just reads from the `/proc/stat` file and spits out a file with timestamps and cpu utilizations.

See in the screenshot as an example. The 208 etc. means 2.08%
<img width="1162" alt="Screenshot 2023-06-15 at 7 21 01 PM" src="https://github.com/green-coding-berlin/green-metrics-tool/assets/250671/31614972-d69c-454c-8865-dfeca365ea49">

So in short: The GMT currently lacks this functionality for macOS but it could be really easily implemented I guess.
Would you be interested in making a PR? 

What needs to be done in the PR is a small piece of code, either C-code / Python or bash or similar, that reads the current CPU utilization from the macOS subsystem. 
We pull that in as metric provider plugin and do the refactoring to make the reporter work from our side.

I am also pulling @ribalba for ideas to provide some guidance

**Comment by ArneTR** [comment]
I removed my earlier comment as I confused your setup with a macOS setup and thus had a wrong interpretation.
I made a separate post for the problem I confused it with here: https://github.com/green-coding-berlin/green-metrics-tool/issues/330

Since you are on a linux box and have the procfs reporter active you should get a metric here for the SDIA reporter.

I just made a demo run on one of our boxes and the data indeed does not show up. This is clearly a bug!
XGBoost however works fine.
And the SDIA data actually shows up in the detail graphs further down. But not as cumulative metric box.

Will post a hotfix asap

**Comment by ArneTR** [comment]
PR is in dev and tests work all fine. 

Are you on the dev branch? If so, can you pull and retry? 

No change of the DB is needed, but a restart of the docker containers from the Green Metrics Tool itself under `/docker`

**Comment by ArneTR** [comment]
Note: The old measurements and display will not get updated. you need to make a new run.

**Comment by marvil07-adapt** [comment]
@ArneTR, thanks for the fix!

I verified that the machine energy is being now reported correctly using sdia! :tada:

![Screenshot 2023-06-16 at 11-34-49 Green Metrics Tool](https://github.com/green-coding-berlin/green-metrics-tool/assets/75342537/7397462a-f5cf-4d95-8401-163aedf188bf)

It may be a different issue, but the one missing bit there is the _Machine CO2 (manufacturing)_ value.
**Q**. Is there something else I need configure for that bit to appear?

Also, in the process of jumping to the development branch, I needed to re-install python dependencies.
I added a PR to the documentation to mention it at https://github.com/green-coding-berlin/documentation/pull/31.

**Comment by ArneTR** [comment]
1. Thx, for the PR. I commented there separately.

2. The Machine CO_2 is atm only a placeholder. Sadly we are a bit behind with the documentation on the new 0.15 version but will hope to get that all up during the next week.

The process how we expect to implement it though is a time attribution for the machine CO2. If you are interested in that idea we have written it down in an article here: https://publication2023.bits-und-baeume.org/#40

@djesic For the documentation next week: Can you also please write down that the machine CO2 is a placeholder atm and how we plan to integrate it with the wording from the post here and parts of our article please

**Comment by marvil07-adapt** [comment]
@ArneTR Thanks for the extra feedback, and sharing the planned approach :+1: 
I appreciate it a lot!

Given the original report here was about sdia not working, and that is not verified to be fixed on mainline, I am closing this issue.
If there is something else pending I may not be aware, please re-open as needed.
